File: examples/injectable-twitter-server/java/src/main/java/com/twitter/finatra/example/ExampleTwitterServer.java
package com.twitter.finatra.example;

import java.util.Collection;

import com.google.common.collect.ImmutableList;
import com.google.inject.Module;

import com.twitter.app.Flaggable;
import com.twitter.inject.server.AbstractTwitterServer;

public class ExampleTwitterServer extends AbstractTwitterServer {

  public ExampleTwitterServer() {
    createFlag(
        /* name = */ "subscriber.max.read",
        /* default = */ 5,
        /* help = */ "Subscriber Max read",
        /* flaggable = */ Flaggable.ofJavaInteger());
  }

  @Override
  public Collection<Module> javaModules() {
    return ImmutableList.of(QueueModule.getInstance());
  }

  @Override
  public void setup() {
    // Create/start a pub-sub component and add it to the list of Awaitables, e.g., await(component)
    // It is important to remember to NOT BLOCK this method. We add the publisher and subscriber to
    // the list of awaitables to block on as we want to ensure we entangle them together such that
    // if either exits, the entire server will exit.

    Publisher publisher = injector().instance(Publisher.class);
    await(publisher);

    Subscriber subscriber = injector().instance(Subscriber.class);
    await(subscriber);

    closeOnExit(publisher); // close publisher first to stop writing anything to the queue.
    closeOnExitLast(subscriber); // close subscriber last to ensure everything published is read.
  }

  @Override
  public void start() {
    Publisher publisher = injector().instance(Publisher.class);
    Subscriber subscriber = injector().instance(Subscriber.class);

    subscriber.start(); // start subscriber first in order to not miss anything published.

    // start publisher last to ensure we have a subscriber listening before publishing any item.
    publisher.start();
  }


  /**
   * Callback method run before {@code TwitterServer.postWarmup}, used for performing warm up of this server.
   * Override, but do not call {@code super.warmup()} as you will trigger a lint rule violation.
   *
   * Any exceptions thrown in this method will result in the app exiting.
   */
  @Override
  public void warmup() {
    // do nothing
  }
}

File: examples/injectable-twitter-server/scala/src/main/scala/com/twitter/finatra/example/ExampleTwitterServer.scala
package com.twitter.finatra.example

import com.google.inject.Module
import com.twitter.inject.Logging
import com.twitter.inject.server.TwitterServer

object ExampleTwitterServerMain extends ExampleTwitterServer

class ExampleTwitterServer extends TwitterServer with Logging {
  flag[Int]("subscriber.max.read", 5, "Subscriber Max read")

  override val modules: Seq[Module] = Seq(QueueModule)

  override protected def setup(): Unit = {
    // Create/start a pub-sub component and add it to the list of Awaitables, e.g., await(component)
    // It is important to remember to NOT BLOCK this method. We add the publisher and subscriber to
    // the list of awaitables to block on as we want to ensure we entangle them together such that
    // if either exits, the entire server will exit.

    val publisher = injector.instance[Publisher]
    await(publisher)

    val subscriber = injector.instance[Subscriber]
    await(subscriber)

    closeOnExit(publisher) // close publisher first to stop writing anything to the queue
    closeOnExitLast(subscriber) // close subscriber last to ensure everything published is read.
  }

  override protected def start(): Unit = {
    val publisher = injector.instance[Publisher]
    val subscriber = injector.instance[Subscriber]

    subscriber.start() // start subscriber first in order to not miss anything published.

    // start publisher last to ensure we have a subscriber listening before publishing any item.
    publisher.start()
  }

  /**
   * Callback method run before [[TwitterServer.postWarmup]], used for performing warm up of this server.
   * Override, but do not call `super.warmup()` as you will trigger a lint rule violation.
   *
   * Any exceptions thrown in this method will result in the app exiting.
   */
  override protected def warmup(): Unit = { /* do nothing */ }
}
